src/comm.jl

```diff
@@ -1,264 +1,1531 @@
+
+            #       MPI_Type_vector(xferFields, 1, (fieldData[1] - fieldData[0]),
+            #                       msgTypePencil, &msgTypeCol) ;
+            #       MPI_Type_commit(&msgTypeCol) ;
+            #       packCol = false ;
+            #    }
+            #    else {
+            #       msgTypeCol = baseType ;
+            #       packCol = true ;
+            #    }
+            # }
+            @assert !ALLOW_UNPACKED_COL
+            packCol = true
+
+            if colMin
+               # contiguous memory
+               if packCol
+                  srcOffset = 0
+                  offset = pmsg * maxPlaneComm
+                  for field in fields
+                     for i in 0:(dz-1)
+                        for j in 0:(dy-1)
+                           # INDEXING +1 missign somewhere
+                           domain.commDataSend[offset+i*dx+j] = field[srcOffset+i*dx*dy+j*dx]
+                        end
+                     end
+                     offset += sendCount
+                  end
+                  # destAddr = &domain->commDataSend[pmsg * maxPlaneComm] ;
+                  # for (Index_t fi=0; fi<xferFields; ++fi) {
+                  #    for (Index_t i=0; i<dz; ++i) {
+                  #       Real_t *srcAddr = &fieldData[fi][i*dx*dy] ;
+                  #       for (Index_t j=0; j<dy; ++j) {
+                  #          destAddr[i*dy + j] = srcAddr[j*dx] ;
+                  #       }
+                  #    }
+                  #    destAddr += sendCount ;
+                  # }
+                  idx = pmsg * maxPlaneComm + 1
+                  src = view(domain.commDataSend, idx:(idx+(xferFields * sendCount)))
+               else
+                  error("unimplemented")
+               end
+               otherRank = myRank - 1
+               req = MPI.Isend(src, otherRank, msgType + myRank, domain.comm)
+               domain.sendRequest[pmsg+1] = req
+               pmsg += 1
+            end
+
+            if colMax && doSend
+               # contiguous memory
+               if packCol
+                  srcOffset = dx - 1
+                  offset = pmsg * maxPlaneComm
+                  for field in fields
+                     for i in 0:(dz-1)
+                        for j in 0:(dy-1)
+                           # INDEXING +1 missign somewhere
+                           domain.commDataSend[offset+i*dy+j] = field[srcOffset+j*dx]
+                        end
+                     end
+                     offset += sendCount
+                  end
+                  idx = pmsg * maxPlaneComm + 1
+                  src = view(domain.commDataSend, idx:(idx+(xferFields * sendCount)))
+               else
+                  error("unimplemented")
+               end
+               otherRank = myRank + 1
+               req = MPI.Isend(src, otherRank, msgType + myRank, domain.comm)
+               domain.sendRequest[pmsg+1] = req
+               pmsg += 1
+            end
+
```

Comment:
This plane only down should be a top level if.

Reply:
I think I corrected that now.